src/cascadia/TerminalCore/Terminal.cpp

```diff
@@ -1584,6 +1584,57 @@ til::point Terminal::GetViewportRelativeCursorPosition() const noexcept
     return absoluteCursorPosition - viewport.Origin();
 }
 
+void Terminal::PreviewText(std::wstring_view input)
+{
+    // Our suggestion text is default-on-default, in italics.
+    static constexpr TextAttribute previewAttrs{ CharacterAttributes::Italics, TextColor{}, TextColor{}, 0u, TextColor{} };
+
+    auto lock = LockForWriting();
+    if (input.empty())
+    {
+        snippetPreview.text = L"";
+        snippetPreview.attributes.clear();
+        _activeBuffer().NotifyPaintFrame();
+        return;
+    }
+
+    // When we're previewing suggestions, they might be preceded with DEL
+    // characters to backspace off the old command.
+    //
+    // But also, in the case of something like pwsh, there might be MORE "ghost"
+    // text in the buffer _after_ the commandline.
+    //
+    // We need to trim off the leading DELs, then pad out the rest of the line
+    // to cover any other ghost text.
+    std::wstring_view view{ input };
+    // Where do the DELs end?
+    const auto strBegin = view.find_first_not_of(L"\x7f");
+    if (strBegin != std::wstring::npos)
+    {
+        // Trim them off.
+        view = view.substr(strBegin);
+    }
+    // How many spaces do we need, so that the preview exactly covers the entire
+    // prompt, all the way to the end of the viewport?
+    const auto bufferWidth = _GetMutableViewport().Width();
+    const auto cursorX = _activeBuffer().GetCursor().GetPosition().x;
+    const auto expectedLenTillEnd = strBegin + (bufferWidth - cursorX);
+    std::wstring preview{ view };
+    const auto originalSize{ preview.size() };
+    if (expectedLenTillEnd > originalSize)
+    {
+        // pad it out
+        preview.insert(originalSize, expectedLenTillEnd - originalSize, L' ');
+    }

```

Comment:
(This is not compatible with wide glyphs or much of Unicode. We talked about it internally and agreed to fix it later when it's not an experimental feature. FWIW the `\x7f`-search is also not that good of a representation and we may want to replace it with something else as well. I proposed sending an Esc key press, because it clears the prompt in most shells.)

Reply:
Fair. Is there an issue tracking this?

Reply:
FWIW I didn't do the escape thing, because I wanted to be able to go back and overwrite the currently typed command, and that was hard to do if all I had was an escape. 

For example, for the commandline `gco`, and the suggestion `git commit`, the thing we generate right now is `\x7f\x7f\x7fgit commit` - the three backspaces letting us know how long that input originally was. 

I'll file something to specifically update the few places that all do the same thing here.

Reply:
resolving for now in favor of #17419